crates/orderbook/src/app_data.rs

```diff
@@ -0,0 +1,74 @@
+use {
+    crate::database::{app_data::InsertError, Postgres},
+    model::app_id::AppDataHash,
+    shared::app_data,
+};
+
+/// CoW Protocol API app-data registry.
+pub struct Registry {
+    validator: app_data::Validator,
+    database: Postgres,
+}
+
+impl Registry {
+    /// Creates a new instance of an app-data registry.
+    pub fn new(validator: app_data::Validator, database: Postgres) -> Self {
+        Self {
+            validator,
+            database,
+        }
+    }
+
+    /// Returns the size limit, in bytes, of an app-data document.
+    pub fn size_limit(&self) -> usize {
+        self.validator.size_limit()
+    }
+
+    /// Saves an app-data document matching the specified app-data hash to the
+    /// database.
+    pub async fn save(&self, hash: AppDataHash, document: &[u8]) -> Result<(), SaveError> {
+        let validated = self
+            .validator
+            .validate(document)
+            .map_err(SaveError::Invalid)?;
+        if hash != validated.hash {
+            return Err(SaveError::HashMismatch {
+                expected: hash,
+                computed: validated.hash,
+            });
+        }
+
+        self.database
+            .insert_full_app_data(&validated.hash, &validated.document)
+            .await?;
+
+        Ok(())
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum SaveError {
+    #[error("app data is invalid: {0}")]
+    Invalid(anyhow::Error),
+    #[error("app data already exists")]
+    Duplicate,
```

Comment:
Should this be an error?

Reply:
I thought the API was easier to deal with if it was.

The alternative would be to have a:

```rust
enum Saved {
    New,
    AlreadyExists,
}
```

And return a

```rust
Result<Saved, SaveError>
```

I originally went this way but the API was a bit clunkier. Happy to change it if you feel this is more correct.

Reply:
I have a weak preference for not erroring if the data is already there. The point of this route is to make this app data available in the database. If it is already there then that isn't an error to the consumer of this API. The consumer doesn't need to know whether the app data was freshly saved or was already there.

Reply:
Actually, it ended up looking "nicer" IMO. Thanks for the suggestion!

275b21082365c6d966dcb30ee80688e9f79acf86